core/src/main/scala/org/broadinstitute/dsp/workbench/welder/package.scala

```diff
@@ -82,14 +84,35 @@ package object welder {
   //This string will be hashed when stored in GCS metadata
   def lockedByString(bucketName: GcsBucketName, ownerEmail: WorkbenchEmail): String = bucketName.value + ":" + ownerEmail.value
 
-  def hashString(metadata: String): IO[HashedLockedBy] = IO {
+  def hashString(metadata: String): Either[Throwable, HashedLockedBy] = Either.catchNonFatal {
     HashedLockedBy(String.format("%032x", new BigInteger(1, MessageDigest.getInstance("SHA-256").digest(metadata.getBytes("UTF-8")))))
   }
 
-  type StorageLinksCache = Ref[IO, Map[Path, StorageLink]]
-  type MetadataCache = Ref[IO, Map[Path, AdaptedGcsMetadataCache]]
+  type StorageLinksCache = Ref[IO, Map[RelativePath, StorageLink]]
+  type MetadataCache = Ref[IO, Map[RelativePath, AdaptedGcsMetadataCache]]
 
   val gcpObjectType = "text/plain"
 
+  def cachedResource[A, B: Decoder: Encoder](path: Path, blockingEc: ExecutionContext, toTuple: B => List[(A, B)])(
+    implicit logger: Logger[IO], cs: ContextShift[IO]
+  ): Stream[IO, Ref[IO, Map[A, B]]] =
+    for {
+      cached <- readJsonFileToA[IO, List[B]](path).map(ls => ls.flatMap(b => toTuple(b)).toMap).handleErrorWith { error =>
+        Stream.eval(logger.info(s"$path not found")) >> Stream.emit(Map.empty[A, B]).covary[IO]
+      }
+      ref <- Stream.resource(
+        Resource.make(Ref.of[IO, Map[A, B]](cached))(
+          ref =>
```

Comment:
Having trouble sorting out the streams -- when does this release function run? Is it after each usage of the cached resource, after every API request, or at the end of the app (e.g. when welder shuts down)? I want to make sure we're reliably flushing caches to disk and not losing any data if for example the VM gets paused.

Reply:
Every time the app starts up, we load the file into memory. Release should be guaranteed to run when app shuts down.

Reply:
Does the app shut down cleanly though when the cluster is stopped? Might be worth a test.

Reply:
Sure...https://broadworkbench.atlassian.net/browse/IA-1107